src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs

```diff
@@ -547,51 +549,61 @@ private static void UpdateOptionsForNextIteration(ref ChatOptions options, strin
+    /// <param name="isStreaming">Whether the function calls are being processed in a streaming context.</param>
     /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests.</param>
     /// <returns>A value indicating how the caller should proceed.</returns>
-    private async Task<(bool ShouldTerminate, int NewConsecutiveErrorCount, IList<ChatMessage> MessagesAdded)> ProcessFunctionCallsAsync(
-        List<ChatMessage> messages, ChatOptions options, List<FunctionCallContent> functionCallContents, int iteration, int consecutiveErrorCount, CancellationToken cancellationToken)
+    protected virtual async Task<(bool ShouldTerminate, int NewConsecutiveErrorCount, IList<ChatMessage> MessagesAdded)> ProcessFunctionCallsAsync(
+        ChatResponse response, IList<ChatMessage> messages, ChatOptions options, IList<FunctionCallContent> functionCallContents, int iteration, int consecutiveErrorCount,
+        bool isStreaming, CancellationToken cancellationToken)
     {
+        _ = Throw.IfNull(messages);
+        _ = Throw.IfNull(functionCallContents);
+
         // We must add a response for every tool call, regardless of whether we successfully executed it or not.
         // If we successfully execute it, we'll add the result. If we don't, we'll add an error.
 
-        Debug.Assert(functionCallContents.Count > 0, "Expecteded at least one function call.");
+        Debug.Assert(functionCallContents.Count > 0, "Expected at least one function call.");
 
         var captureCurrentIterationExceptions = consecutiveErrorCount < _maximumConsecutiveErrorsPerRequest;
 
         // Process all functions. If there's more than one and concurrent invocation is enabled, do so in parallel.
         if (functionCallContents.Count == 1)
         {
             FunctionInvocationResult result = await ProcessFunctionCallAsync(
-                messages, options, functionCallContents, iteration, 0, captureCurrentIterationExceptions, cancellationToken);
+                messages, options, functionCallContents, iteration, 0, captureCurrentIterationExceptions, isStreaming, cancellationToken);
 
-            IList<ChatMessage> added = CreateResponseMessages([result]);
-            ThrowIfNoFunctionResultsAdded(added);
-            UpdateConsecutiveErrorCountOrThrow(added, ref consecutiveErrorCount);
+            IList<ChatMessage> addedMessages = CreateResponseMessages([result]);
+            ThrowIfNoFunctionResultsAdded(addedMessages);
+            UpdateConsecutiveErrorCountOrThrow(addedMessages, ref consecutiveErrorCount);
 
-            messages.AddRange(added);
-            return (result.ShouldTerminate, consecutiveErrorCount, added);
+            foreach (var addedMessage in addedMessages)
+            {
+                messages.Add(addedMessage);
+            }
```

Comment:
Why do this instead of `AddRange`?

Reply:
As this method is not private anymore, `IList` doesn't have `AddRange`.

Reply:
The 99.9% case will be `List<T>`, and AddRange is significantly more efficient than adding elements one at a time. If we have to accept `IList`, we should still use AddRange if the list type tests as `List<T>`.